### DISCUSSION TITLE
Best way to develop / test whole z2m ecosystem

### DISCUSSION THREAD

**Discussion by vaizki** [Koenkk/zigbee2mqtt#11930] [General]
So I have already contributed a few new devices but it was a pretty painful process for me to develop (not really a node / js dev). So I'm looking for guidance on how to smoothly develop & contribute to all the components.

My current idea / setup:
- Fork all repos: zigbee2mqtt, zigbee-herdsman, zigbee-herdsman-converters and zigbee2mqtt-frontend
- Clone all repos onto my home automation box
- Run `sudo npm link` in all repos except zigbee2mqtt
- ...?

Now this is where I get a bit lost.. 
- With all the inter-depencies and `npm ci` installing them, how should I uninstall / replace the dependency package with the `npm link` version from local filesystem?
- How and in what order should I build the repos to produce a zigbee2mqtt that I can `npm start` in my own "production" use (read: home assistant).

I have the basics of contributing down, creating branches & PRs, rebasing etc but just not this npm + typescript voodoo.

**Comment by Koenkk** [comment]
My personal development setup:
- create a directory (e.g. `zigbee`)
- in this directory clone all repos
- execute `npm ci` in all cloned repos
- `cd zigbee2mqtt/node_modules`
- for each dependency you want to replace: `rm -rf zigbee-herdsman && ln -s ../../zigbee-herdsman zigbee-herdsman`
- run zigbee2mqtt and zigbee-herdsman in watch mode such that a rebuild is triggered every-time you change the `*.ts` files:
  - `cd zigbee-herdsman && npm start`
  - `cd zigbee2mqtt && npm run build-watch`
